mediatek: Add support for device zbtlink zbt-z8103ax-c

joelinux60 opened the pull request:
Specifications:

SoC: MediaTek MT7981B
RAM: 256MiB
Flash: Winbond SPI-NAND 128 MiB
Switch: 1 WAN, 3 LAN (Gigabit) MediaTek MT7531
Buttons: Reset, Mesh
Power: DC 12V 1A
WiFi: MT7981B 2.4Ghz & 5.8Ghz

Led Layout from bottom to top:

- Power
- Mesh (RGB Led, user controllable, default set to OpenWrt Status)
- WLAN 2.4 GHz (user controllable)
- WAN  (user controllable)
- LAN3
- LAN2
- LAN1
- WLAN 5 GHz (Not on panel but blinks through enclosure, user controllable)

Buttons:

- Reset
- Mesh (user controllable, no default function)

Installation:

A. Through U-Boot menu:

1.  Prepare your connecting computer to use a static IP in network 192.168.1.0/24
2.  Power down the router and hold in the Reset button.
3.  While holding in the button power up the router again.
4.  Hold the button in for 10 seconds and then release.
5.  Use your browser to go to 192.168.1.1
6.  If you see a GUI allowing for flashing firmware then you got the right model.
7.  Upload the Factory file.

Note: Recovery GUI it can be used to recover from an incorrect firmware flash.

B. Through OpenWrt Dashboard:
- If your router comes with OpenWrt preinstalled (modified by the seller), you can easily upgrade by going to the dashboard (192.168.1.1) and then navigate to System -> Backup/Flash firmware, then flash the firmware

MAC Addresses were found in Factory partition:

offset 0x4  F8:5E:3C:xx:xx:aa --> Router Label -2
offset 0xa  F8:5E:3C:xx:xx:bb --> Router Label -1
offset 0x24 F8:5E:3C:xx:xx:cc --> Router Label +1
offset 0x2a F8:5E:3C:xx:xx:yy --> printed on Router Label

castiel652 commented:
Looks like the DTS is the same?

joelinux60 commented:
> Looks like the DTS is the same?

Almost the same. First partition on existing device was renamed from all lowercase bl2 to uppercase BL2.

joelinux60 commented:
> > Almost the same. First partition on existing device was renamed from all lowercase bl2 to uppercase BL2.
> 
> Whether the partition name is uppercase or lowercase does not really make a difference. Because some people will change the partition name to lowercase when submitting.

I'm confused about statement of uppercase or lowercase of MTD partition names make no difference. 

I did use firmware selector to build firmware for Zbtlink ZBT-Z8103AX. Did interrupt boot process by pressing and holding RESET button to get U-BOOT menu. From there I did upload Factory Image previously downloaded from firmware selector.
My Z8103AX-C runs into a boot loop and I had to recover means of using the U-BOOT menu.

My model C devices fail to boot while trying to use existing DTC for Zbtlink ZBT-Z8103AX. 

My current understanding is that a DTC has to match case of (Vendor provided) MTD partition names to allow kernel to successfully find root filesystem. This is what I did here, match case of (Vendor provided) MTD partition names.

aiamadeus commented:
> My model C devices fail to boot while trying to use existing DTC for Zbtlink ZBT-Z8103AX.

This is because the size of the ubi partition is different, not the partition name:
https://github.com/openwrt/openwrt/blob/6f33529a1249826f0bbedd880c4c27e4cbe8a0f0/target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts#L194-L197
https://github.com/openwrt/openwrt/blob/e30363374b535ea02623cbea81ae75cb86b38039/target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax-c.dts#L236-L239
So this is a new device, not an already supported one.

hauke commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dtsi:
Please change `target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts` to use this dtsi file too.

hauke commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dtsi:
```diff
@@ -0,0 +1,290 @@
+        eeprom_factory: eeprom@0 {
+		reg = <0x0 0x1000>;
+	};
+};
```
Please add a empty line at the end.

joelinux60 replied:
fixed

hauke commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dtsi:
```diff
@@ -0,0 +1,290 @@
+&wifi {
+	status = "okay";
+	nvmem-cells = <&eeprom_factory>;
+	nvmem-cell-names = "eeprom";
```
Please set the mac addresses here:
```
	band@0 {
		reg = <0>;
		nvmem-cells = <&macaddr_factory_4 0>;
		nvmem-cell-names = "mac-address";
	};

	band@1 {
		reg = <1>;
		nvmem-cells = <&macaddr_factory_4 1>;
		nvmem-cell-names = "mac-address";
	};
```

hauke replied:
With this change the setting in `target/linux/mediatek/filogic/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac` is not needed any more. 

Please check that the mac addresses match what the vendor configured after doing these changes.

hauke commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dtsi:
```diff
@@ -0,0 +1,290 @@
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	macaddr_factory_004: macaddr@4 {
+		reg = <0x004 0x6>;
+	};
```
To set an mac address offset change it like this:
```
					macaddr_factory_4: macaddr@4 {
						reg = <0x4 0x6>;
						compatible = "mac-base";
						#nvmem-cell-cells = <1>;
					};
```

Also remove the leeding 0.

DragonBluep commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax-c.dts:
```diff
@@ -0,0 +1,11 @@
+
+/ {
+	model = "Zbtlink 3000M(WiFi6) ZBT-Z8103AX-C";
+	compatible = "zbtlink,zbt-z8103ax-c", "mediatek,mt7981";
```
```suggestion
	compatible = "zbtlink,zbt-z8103ax-c", "mediatek,mt7981b";
```
There are some changes in 6.12 kernel.

joelinux60 replied:
adopted

DragonBluep commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -2,134 +2,12 @@
 / {
 	model = "Zbtlink ZBT-Z8103AX";
 	compatible = "zbtlink,zbt-z8103ax", "mediatek,mt7981";
 
```
Remove this empty line.

DragonBluep reviewed (commented):
Please rebase and rework this PR. IIRC there are some new changes in `mt7981b-zbtlink-zbt-z8103ax.dts‎` recently.

joelinux60 reviewed (commented):
> ```suggestion
> 	compatible = "zbtlink,zbt-z8103ax-c", "mediatek,mt7981b";
> ```
> There are some changes in 6.12 kernel. 

adopted

joelinux60 commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax-c.dts:
```diff
@@ -0,0 +1,11 @@
+/ {
+	model = "Zbtlink 3000M(WiFi6) ZBT-Z8103AX-C";
+	compatible = "zbtlink,zbt-z8103ax-c", "mediatek,mt7981";
+
```
done

hauke commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dtsi:
Thank for extracting the commit dtsi file.

Please change the order of the commits in this PR, first extract it and then make use of it.

joelinux60 committed:
mediatek: extract DTS include for zbt z8103ax

- Use .dtsi for old z8103ax featuring a 64M Nand
- Prepare .dtsi for model C of z8103ax featuring a 128M Nand

This .dtsi is supposed to match boards labeled as hardware
revision Z8103AX_V01. Model variant D of zbt-z8103ax appears
to use very same board.

DTS include does also

- set wifi mac addresses in &wifi section
- set eth mac addresses in &eth section
- set LAN switch mac addresses in &switch section

All of the above allows to get rid of 11_fix_wifi_mac script

Signed-off-by: Jörg Seitz <github.joeterminal@xoxy.net>
Link: https://github.com/openwrt/openwrt/pull/19823
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

joelinux60 committed:
mediatek: add support for device zbt-z8103ax-c

- adopted MTD partition size increase from 64M to 128M
  -> this allows boot to complete
- moved MAC addressing stuff out of partition definitions
- all MAC addressing stuff is in .dtsi

Specifications:

SoC: MediaTek MT7981B
RAM: 256MiB
Flash: Winbond SPI-NAND 128 MiB
Switch: 1 WAN, 3 LAN (Gigabit) MediaTek MT7531
Buttons: Reset, Mesh
Power: DC 12V 1A
WiFi: MT7981B 2.4Ghz & 5.8Ghz

Led Layout from bottom to top:

    Power
    Mesh (RGB Led, user controllable, default set to OpenWrt Status)
    WLAN 2 GHz (user controllable)
    WAN (user controllable)
    LAN3
    LAN2
    LAN1
    WLAN 5 GHz (Not on front panel but blinks through enclosure,
               user controllable)

Buttons:

    Reset
    Mesh (user controllable, no default function)

Installation:

A. Through U-Boot menu:

 -  Prepare your connecting computer to use a static IP in
	network 192.168.1.0/24
 -  Power down the router and hold in the Reset button.
 -  While holding in the button power up the router again.
 -  Hold the button in for 10 seconds and then release.
 -  Use your browser to go to 192.168.1.1
 -  If you see a GUI allowing for flashing firmware then
	you got the right model.
 -  Upload the sysupgrade file.

Note: Recovery GUI can be used to recover from an incorrect
      firmware flash.

B. Through OpenWrt Dashboard:

    If your router comes with OpenWrt preinstalled
	(modified by the seller), you can easily upgrade
	by going to the dashboard (192.168.1.1) and then
	navigate to System -> Backup/Flash firmware,
	then flash the firmware

MAC Addresses were found in Factory partition:

offset 0x4 F8:5E:3C:xx:xx:aa --> Router Label -2
offset 0xa F8:5E:3C:xx:xx:bb --> Router Label -1
offset 0x24 F8:5E:3C:xx:xx:cc --> Router Label +1
offset 0x2a F8:5E:3C:xx:xx:yy --> printed on Router Label

Signed-off-by: Jörg Seitz <github.joeterminal@xoxy.net>
Link: https://github.com/openwrt/openwrt/pull/19823
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

joelinux60 commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dtsi:
```diff
@@ -0,0 +1,314 @@
+
+	memory {
+		reg = <0 0x40000000 0 0x10000000>;
+	};
```
done

joelinux60 commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dtsi:
```diff
@@ -0,0 +1,314 @@
+	spi_nand@0 {
+		compatible = "spi-nand";
+		#address-cells = <1>;
+		#size-cells = <1>;
```
done

joelinux60 commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dtsi:
```diff
@@ -0,0 +1,314 @@
+		led-failsafe = &led_status_red;
+		led-running = &led_status_green;
+		led-upgrade = &led_status_green;
+		label-mac-device = &gmac1;
```
Thought is that symbol label-mac-device should denote mac printed on router label. Router label mac is the one at offset 0x2a and is assigned to wan interface eth1 labelled gmac1.
Hope these thoughts appear to make sense.

joelinux60 commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dtsi:
```diff
@@ -0,0 +1,303 @@
+		gpios = <&pio 2 GPIO_ACTIVE_HIGH>;
+		hw_algo = "toggle";
+		hw_margin_ms = <1000>;
+	};
```
Going to remove watchdog node. Compile is running.

joelinux60 replied:
Board keeps running with watchdog node removed from .dtsi file. No unusual reboots.
Thank you for reminding me to verify watchdog definitions. It turns out that board hasn't any hardware watchdog.
